### PR TITLE
ingress: Bypass service count check

### PR DESCRIPTION
**Pull request by sayboras**
### Description

As L7 LB service is having dummy endpoint (e.g. 192.192.192.192), the
in-cluster traffic will be dropped due to no healthy backend. This commit
is to make sure that the proper exclusion rule is done.

Relates: https://github.com/cilium/cilium/pull/21871
Fixes: 38959d4c07fcbf6be361be253d8187e1c3732553

Signed-off-by: Tam Mach <tam.mach@cilium.io>

Thanks to @julianwiedmann and @brb 

```release-note
Fix incorrectly dropping in-cluster traffic for L7 ingress resources
```

### CODE REVIEW

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -322,7 +322,12 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 	if (!svc)
 		return -ENXIO;
 	if (svc->count == 0)
+#if defined(ENABLE_L7_LB)
+		if (!lb4_svc_is_l7loadbalancer(svc))
+			return -EHOSTUNREACH;
```
Comment (borkmann) [inline bpf/bpf_sock.c]: Pls also add a check for IPv6 path.

**Review by nbusseneau — APPROVED** [review]

**Review by brb — COMMENTED** [review]
LGTM, thanks. Could you also add to the fix commit msg `Fixes: 38959d4c07 ("bpf: Block service connections if no backend is available")`?

**Comment by sayboras** [comment]
> LGTM, thanks. Could you also add to the fix commit msg `Fixes: 38959d4c07 ("bpf: Block service connections if no backend is available")`?

Good point, done and thanks.

**Comment by sayboras** [comment]
/test-1.26-net-next

File: bpf/bpf_sock.c [hunk 2]
```diff
@@ -322,7 +322,12 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+#else
 		return -EHOSTUNREACH;
+#endif
```
Comment (julianwiedmann) [inline bpf/bpf_sock.c]:
```suggestion
		if (!lb4_svc_is_l7loadbalancer(svc))
			return -EHOSTUNREACH;
```

It's ok to call `lb4_svc_is_l7loadbalancer()` when L7 LB is disabled. Same for IPv6.

**Review by julianwiedmann — COMMENTED** [review]
Thanks for the fix Tam!

If you need to push once more, mind clarifying the patch subject - something like `"bpf: socketLB: consider L7 LB when checking svc backend count"` ? :pray:

I wonder whether `__sock4_xlate_rev()` needs the same fix :thinking:.

File: bpf/bpf_sock.c (hunk 2 above)
Comment (sayboras) [reply bpf/bpf_sock.c]: ah you are right, lb4_svc_is_l7loadbalancer already checks if ENABLE_L7_LB is enabled.

**Review by gentoo-root — APPROVED** [review]

**Review by julianwiedmann — APPROVED** [review]

**Comment by sayboras** [comment]
/test-1.16-4.19

**Comment by margamanterola** [comment]
I've updated the release note to try to make it a bit more obvious for users what this PR is fixing ("ingress: Bypass service count check" doesn't tell users what the impact of the bugfix is for them). Please check if my description correctly represents the problem that this PR is fixing. Thanks!

**Pull request merged**